<img src="http://imgur.com/1ZcRyrc.png" style="float: left; margin: 20px; height: 55px">
## Pandas Concatenation Lab

_Authors: Riley Dallas (ATX), Dave Yerrington (SF)_

Here's how you import a dataset from Scikit Learn:

In [72]:
from sklearn.datasets import load_iris
import pandas as pd

Scikit Learn datasets are actually functions that return an object containing the data we need. 

#### In the cell below, call the `load_iris()` function and set the result to a variable called `data`.

In [73]:
data = load_iris()

In [74]:
data

{'data': array([[5.1, 3.5, 1.4, 0.2],
        [4.9, 3. , 1.4, 0.2],
        [4.7, 3.2, 1.3, 0.2],
        [4.6, 3.1, 1.5, 0.2],
        [5. , 3.6, 1.4, 0.2],
        [5.4, 3.9, 1.7, 0.4],
        [4.6, 3.4, 1.4, 0.3],
        [5. , 3.4, 1.5, 0.2],
        [4.4, 2.9, 1.4, 0.2],
        [4.9, 3.1, 1.5, 0.1],
        [5.4, 3.7, 1.5, 0.2],
        [4.8, 3.4, 1.6, 0.2],
        [4.8, 3. , 1.4, 0.1],
        [4.3, 3. , 1.1, 0.1],
        [5.8, 4. , 1.2, 0.2],
        [5.7, 4.4, 1.5, 0.4],
        [5.4, 3.9, 1.3, 0.4],
        [5.1, 3.5, 1.4, 0.3],
        [5.7, 3.8, 1.7, 0.3],
        [5.1, 3.8, 1.5, 0.3],
        [5.4, 3.4, 1.7, 0.2],
        [5.1, 3.7, 1.5, 0.4],
        [4.6, 3.6, 1. , 0.2],
        [5.1, 3.3, 1.7, 0.5],
        [4.8, 3.4, 1.9, 0.2],
        [5. , 3. , 1.6, 0.2],
        [5. , 3.4, 1.6, 0.4],
        [5.2, 3.5, 1.5, 0.2],
        [5.2, 3.4, 1.4, 0.2],
        [4.7, 3.2, 1.6, 0.2],
        [4.8, 3.1, 1.6, 0.2],
        [5.4, 3.4, 1.5, 0.4],
        [5.2, 4.1, 1.5, 0.1],
  

In machine learning, the column we're trying to predict is usually called the **target** (or **label**). 

#### To see the targets for our dataset, call `data['target']` in the cell below.

> Also, you can use `data.keys()` to see a list of elements available inside the `data` object.
>```python
> dict_keys(['data', 'target', 'target_names', 'DESCR', 'feature_names'])
>```

In [75]:
data['target']

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

In the case of the iris dataset, the target is the particular species of iris flower. Because machine learning requires our features and target to be numbers, the species are encoded as 0, 1, 2. These indices correspond to the labeled species on `data.target_names`

#### Call `data['target_names']` to see the actual names of the three species

In [76]:
data['target_names']

array(['setosa', 'versicolor', 'virginica'], dtype='<U10')

#### Get the length of `data['target']` to see how many flowers are in this dataset.

In [77]:
len(data['target'])

150

#### Use `data['target']` to create your first pandas DataFrame, which is just a single column (`"species"`).

In [78]:
df1 = pd.DataFrame(data['target'], columns=['species'])

The features for this dataset are on `data['data']`. There are 150 rows, one for each flower and 4 columns, one for each of the following features:
1. sepal length (cm)
2. sepal width (cm)
3. petal length (cm)
4. petal width (cm)

#### Output `data['data']` in the cell below to see the features for this dataset

In [79]:
data['data']

array([[5.1, 3.5, 1.4, 0.2],
       [4.9, 3. , 1.4, 0.2],
       [4.7, 3.2, 1.3, 0.2],
       [4.6, 3.1, 1.5, 0.2],
       [5. , 3.6, 1.4, 0.2],
       [5.4, 3.9, 1.7, 0.4],
       [4.6, 3.4, 1.4, 0.3],
       [5. , 3.4, 1.5, 0.2],
       [4.4, 2.9, 1.4, 0.2],
       [4.9, 3.1, 1.5, 0.1],
       [5.4, 3.7, 1.5, 0.2],
       [4.8, 3.4, 1.6, 0.2],
       [4.8, 3. , 1.4, 0.1],
       [4.3, 3. , 1.1, 0.1],
       [5.8, 4. , 1.2, 0.2],
       [5.7, 4.4, 1.5, 0.4],
       [5.4, 3.9, 1.3, 0.4],
       [5.1, 3.5, 1.4, 0.3],
       [5.7, 3.8, 1.7, 0.3],
       [5.1, 3.8, 1.5, 0.3],
       [5.4, 3.4, 1.7, 0.2],
       [5.1, 3.7, 1.5, 0.4],
       [4.6, 3.6, 1. , 0.2],
       [5.1, 3.3, 1.7, 0.5],
       [4.8, 3.4, 1.9, 0.2],
       [5. , 3. , 1.6, 0.2],
       [5. , 3.4, 1.6, 0.4],
       [5.2, 3.5, 1.5, 0.2],
       [5.2, 3.4, 1.4, 0.2],
       [4.7, 3.2, 1.6, 0.2],
       [4.8, 3.1, 1.6, 0.2],
       [5.4, 3.4, 1.5, 0.4],
       [5.2, 4.1, 1.5, 0.1],
       [5.5, 4.2, 1.4, 0.2],
       [4.9, 3

#### Create a `features` DataFrame in pandas, using `data['data']` and `data['feature_names']` 
> You can use `pd.DataFrame()` to create a DataFrame, passing the `columns` parameter using `data['feature_names']`

In [80]:
df2 = pd.DataFrame(data['data'], columns=(data['feature_names']))

#### Finally, use `pd.concat` to combine the two dataframes.
> Review / research the documentation on the useage of this function.

In [82]:
pd.concat([df1, df2], join='outer', axis=1)

species  sepal length (cm)  sepal width (cm)  petal length (cm)  \
0          0                5.1               3.5                1.4   
1          0                4.9               3.0                1.4   
2          0                4.7               3.2                1.3   
3          0                4.6               3.1                1.5   
4          0                5.0               3.6                1.4   
5          0                5.4               3.9                1.7   
6          0                4.6               3.4                1.4   
7          0                5.0               3.4                1.5   
8          0                4.4               2.9                1.4   
9          0                4.9               3.1                1.5   
10         0                5.4               3.7                1.5   
11         0                4.8               3.4                1.6   
12         0                4.8               3.0                1.4   
13         0                4.3               3.0                1.1   
14         0                5.8               4.0                1.2   
15         0                5.7               4.4                1.5   
16         0                5.4               3.9                1.3   
17         0                5.1               3.5                1.4   
18         0                5.7               3.8                1.7   
19         0                5.1               3.8                1.5   
20         0                5.4               3.4                1.7   
21         0                5.1               3.7                1.5   
22         0                4.6               3.6                1.0   
23         0                5.1               3.3                1.7   
24         0                4.8               3.4                1.9   
25         0                5.0               3.0                1.6   
26         0                5.0               3.4                1.6   
27         0                5.2               3.5                1.5   
28         0                5.2               3.4                1.4   
29         0                4.7               3.2                1.6   
..       ...                ...               ...                ...   
120        2                6.9               3.2                5.7   
121        2                5.6               2.8                4.9   
122        2                7.7               2.8                6.7   
123        2                6.3               2.7                4.9   
124        2                6.7               3.3                5.7   
125        2                7.2               3.2                6.0   
126        2                6.2               2.8                4.8   
127        2                6.1               3.0                4.9   
128        2                6.4               2.8                5.6   
129        2                7.2               3.0                5.8   
130        2                7.4               2.8                6.1   
131        2                7.9               3.8                6.4   
132        2                6.4               2.8                5.6   
133        2                6.3               2.8                5.1   
134        2                6.1               2.6                5.6   
135        2                7.7               3.0                6.1   
136        2                6.3               3.4                5.6   
137        2                6.4               3.1                5.5   
138        2                6.0               3.0                4.8   
139        2                6.9               3.1                5.4   
140        2                6.7               3.1                5.6   
141        2                6.9               3.1                5.1   
142        2                5.8               2.7                5.1   
143        2                6.8              